### ISSUE TITLE
Determine future branching strategy

### ISSUE THREAD

**Issue by kotfu** [python-cmd2/cmd2#713]
When we get to 1.0.0 and are ready to really do semantic versioning, our current branch/merge approach will not be sufficient. After 1.0.0 is tagged, we could have potentially three active development branches:

- 1.0.1
- 1.1.0
- 2.0.0

We will need to rethink how we use **master**. Maybe we need to only merge to master when we cut a release, so **master** HEAD contains the most "recent" (according to semver) release. We probably will also need to start using a **develop** branch as a place to integrated all the hotfix, minor (i.e. 1.x.0) and major (i.e. x.0.0) branches.

We are not the first people to have to solve this problem, after a little digging, I found
[Hendrick Prinsloo's description of his approach](https://medium.com/@HendrikPrinsZA/gitflow-with-semver-dde625429aca). I think this is a reasonably sound method, and propose that we adopt it.

Opinions and debate welcome.

**Comment by tleonhardt** [comment]
I was thinking of keeping it simple where we continue to use **master** as the development branch where we are making backwards compatible bug fixes for the next patch release (change in 3rd digit - e.g. for the 1.0.1 release if we just released 1.0.0.

Then when we want to add new functionality in a backwards-compatible manner, we create a branch for the next minor version, e.g. `1.1.0`.  And when we want to make incompatible API changes, we create a branch for the next major version, e.g. `2.0.0`.  

In this scenario, changes made to lower version get merged to higher branches - i.e. changes to **master** would get merged to **1.1.0** and **2.0.0**.  Then when we are ready to release `1.1.0`, we merge the **1.1.0** branch into **master**, tag it, do a release and delete that branch.

This branching strategy is simpler than the full-blown GitFlow.  GitFlow is designed for a commercial scenario where you need to support customers with multiple versions of your software and apply hot fixes to numerous major/minor versions simultaneously.  The strategy I'm outlining here only supports applying bug fixes to a single minor version (the current one).  As long as we think that is reasonable, I'd prefer to keep it as simple as possible.

But if you guys think we would benefit from switching to a GitFlow approach, then I'd be happy to go that route.

**Comment by kotfu** [comment]
I like your proposal @tleonhardt, but it seems to me that there is only one substantial difference between your recommendation and git flow: git flow has long-running **devel** and **master** branches, you are advocating for only **master**. 

The original [git flow model](https://nvie.com/posts/a-successful-git-branching-model/) can't really handle hot fixes to prior versions either. It doesn't describe how to handle hot fixes that can't be immediately merged into **master**. In some software implementations of the git flow concept, they have the idea of a [support branch](https://gitversion.readthedocs.io/en/latest/git-branching-strategies/gitflow-examples/#support-branches). These support branches are semi-long-lived, made from some long-ago tag in **master**. The support branch gets bug fixes merged into it, and releases cut from it, until that version is no longer supported.

If we are only going to support the "current" version, then git flow could be as simple as two branches: **devel** and **master**. Every time we are ready for a new release, we merge **devel** into **master** and tag it. We finish the 1.x.0 releases before we start on the 2.0.0 release, and we don't need any release branches. 

Whether we have **devel** and **master**, or just **master**, if we intend to have development for 1.x.0 and 2.0.0 happening at the same time, we will need one additional branch, named something like **release-1.x.0**. Version 2.0.0 features get committed to **devel** (or **master** if we aren't using **devel**).

So what if I reframe my original question. Instead of asking whether we adopt git flow, what if I ask:

Should we continue with a single long-lived **master** branch, which gets tagged when we make a release, and into which we merge PR's when they are approved?

or

Should we add a long-lived **develop** branch, into which we merge all of our PR's? When we cut a release, we merge **develop** into **master** and tag it. **master** HEAD always contains the code of the most recent version available at PyPI.


P.S. I must admit my desire for a **develop** branch is somewhat selfish. My git client of choice, Tower, has built in support for git flow, with one click creation of feature, hot fix, and release branches. It also conveniently merges feature branches and deletes them when done. I find it terribly convenient.

**Comment by tleonhardt** [comment]
I'm flexible.  I use [SmartGit](https://www.syntevo.com/smartgit/) as my Git client since it works great on Mac, Linux, and Windows - it has a very configurable GitFlow integration as well as GitHub integration.  

If switching to a particular branching model makes your life easier, I'm OK with that as long as @kmvanbrunt is.  I'd just like to keep it as simple as possible.  

@kotfu How would something like the following work with [Tower](https://www.git-tower.com)?:
* **master** branch used for tagged releases as well as patch version changes prior to a patch release.
* **dev** branch (name negotiable) used for backward-compatible minor release changes 
* **2.0** (or similar) used for backward-compatible major release changes

**Comment by anselor** [comment]
Related to this - I think we should strongly encourage everyone to, before merging a PR, rebase and squash commits down to either 1 commit or a handful of major milestones in a larger feature. It helps to linearize changes and dramatically reduces the cognitive load when trying to look through the history of changes. It makes the git history useful in that it is also acting as a changelog.

**Comment by xNinjaKittyx** [comment]
Just to add to the previous comment, I believe you should be able to select "squash & merge" in Github whenever someone merges a commit.
